src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs

```diff
@@ -62,16 +72,34 @@ public string NewLine
             }
         }
 
-        // Internal for testing.
-        internal CodeWriter Indent(int size)
+        public CodeWriter Indent(int size)
         {
-            if (Length == 0 || this[Length - 1] == '\n')
+            if (size == 0 || (Length != 0 && this[Length - 1] != '\n'))
```

Comment:
This is unrelated to your PR, but indexing in to `StringBuilder` (which is what `this[..]` is doing) is really slow.

Reply:
Hmm, interesting. Maybe we can keep track of the last character somehow, possibly storing it in a field?
(Don't know if it's in the scope of this PR, but would be a good follow up later on.)